domain/network/service/subnet.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	"github.com/juju/juju/core/network"
+)
+
+// SubnetService provides the API for working with subnets.
+type SubnetService struct {
+	st SubnetState
+}
+
+// NewSunbetService returns a new service reference wrapping the input state.
+func NewSubnetService(st SubnetState) *SubnetService {
+	return &SubnetService{
+		st: st,
+	}
+}
+
+// AddSubnet creates and returns a new subnet.
+func (s *SubnetService) AddSubnet(ctx context.Context, args network.SubnetInfo) (network.Id, error) {
+	uuid, err := utils.NewUUID()
+	if err != nil {
+		return "", errors.Annotatef(err, "creating uuid for new subnet with CIDR %q", args.CIDR)
+	}
+
+	if err := s.st.AddSubnet(ctx, uuid.String(), args); err != nil {
+		return "", errors.Trace(err)
+	}
+
+	return network.Id(uuid.String()), nil
+}
+
+// Subnet returns the subnet identified by the input UUID,
+// or an error if it is not found.
+func (s *SubnetService) Subnet(ctx context.Context, uuid string) (*network.SubnetInfo, error) {
+	return s.st.GetSubnet(ctx, uuid)
```

Comment:
We should be tracing/annotating the errors for these, so at least
```go
info, err := s.st.GetSubnet(ctx, uuid)
return info, errors.Trace(err)
```

Reply:
done, see 8c41e31aac